les/server_requests.go

```diff
@@ -0,0 +1,538 @@
+
+			if header = bc.GetHeaderByHash(request.BHash); header == nil {
+				p.Log().Warn("Failed to retrieve header for proof", "hash", request.BHash)
+				p.bumpInvalid()
+				continue
+			}
+			// Refuse to search stale state data in the database since looking for
+			// a non-exist key is kind of expensive.
+			local := bc.CurrentHeader().Number.Uint64()
+			if !backend.ArchiveMode() && header.Number.Uint64()+core.TriesInMemory <= local {
+				p.Log().Debug("Reject stale trie request", "number", header.Number.Uint64(), "head", local)
+				p.bumpInvalid()
+				continue
+			}
+			root = header.Root
+		}
+		// If a header lookup failed (non existent), ignore subsequent requests for the same header
+		if root == (common.Hash{}) {
+			p.bumpInvalid()
+			continue
+		}
+		// Open the account or storage trie for the request
+		statedb := bc.StateCache()
+
+		var trie state.Trie
+		switch len(request.AccKey) {
+		case 0:
+			// No account key specified, open an account trie
+			trie, err = statedb.OpenTrie(root)
+			if trie == nil || err != nil {
+				p.Log().Warn("Failed to open storage trie for proof", "block", header.Number, "hash", header.Hash(), "root", root, "err", err)
+				continue
+			}
+		default:
+			// Account key specified, open a storage trie
+			account, err := getAccount(statedb.TrieDB(), root, common.BytesToHash(request.AccKey))
+			if err != nil {
+				p.Log().Warn("Failed to retrieve account for proof", "block", header.Number, "hash", header.Hash(), "account", common.BytesToHash(request.AccKey), "err", err)
+				p.bumpInvalid()
+				continue
+			}
+			trie, err = statedb.OpenStorageTrie(common.BytesToHash(request.AccKey), account.Root)
+			if trie == nil || err != nil {
+				p.Log().Warn("Failed to open storage trie for proof", "block", header.Number, "hash", header.Hash(), "account", common.BytesToHash(request.AccKey), "root", account.Root, "err", err)
+				continue
+			}
+		}
+		// Prove the user's request from the account or stroage trie
+		if err := trie.Prove(request.Key, request.FromLevel, nodes); err != nil {
+			p.Log().Warn("Failed to prove state request", "block", header.Number, "hash", header.Hash(), "err", err)
+			continue
+		}
+		if nodes.DataSize() >= softResponseLimit {
+			break
+		}
+	}
+	return p.replyProofsV2(reqID, nodes.NodeList())
+}
+
+type GetHelperTrieProofsReq []HelperTrieReq
+
+func (r GetHelperTrieProofsReq) ReqAmount() (uint64, uint64) {
+	return uint64(len(r)), MaxHelperTrieProofsFetch
+}
+
+func (r GetHelperTrieProofsReq) InMetrics(size int64) {
+	miscInHelperTriePacketsMeter.Mark(1)
+	miscInHelperTrieTrafficMeter.Mark(size)
+}
+
+func (r GetHelperTrieProofsReq) OutMetrics(size int64, servingTime time.Duration) {
+	miscOutHelperTriePacketsMeter.Mark(1)
+	miscOutHelperTrieTrafficMeter.Mark(size)
+	miscServingTimeHelperTrieTimer.Update(servingTime)
+}
+
+func (r GetHelperTrieProofsReq) Serve(backend serverBackend, reqID uint64, p *clientPeer, waitOrStop func() bool) *reply {
```

Comment:
Just curious, would it be possible to abstract away the `reqID` from the method handler, and have that handled on the outside? 
Like, `reply.SetReqId(reqID)` ? 

I'm wondering, because for `eth`, when we implement `eth/66`, it would be neat to use a similar handling, but only set the `reqId` on the response in the cases where the protocol is `66`.

Reply:
Well. Not sure it's necessary. `REQID` exists in the protocol all the time.